samples/xamarin-forms/Cinephile/Cinephile.Core/Models/MovieService.cs

```diff
@@ -4,80 +4,69 @@
 
 using System;
 using System.Collections.Generic;
-using System.Globalization;
 using System.Linq;
+using System.Reactive;
 using System.Reactive.Linq;
 using Cinephile.Core.Rest;
-using Cinephile.Core.Rest.Dtos.Movies;
+using DynamicData;
 using Splat;
 
-namespace Cinephile.Core.Model
+namespace Cinephile.Core.Models
 {
     public class MovieService : IMovieService
     {
-        public const int PageSize = 20;
-        const string Language = "en-US";
+        private readonly SourceCache<Movie, int> _internalSourceCache;
+        public IObservableCache<Movie, int> UpcomingMovies => _internalSourceCache;
 
-        const string BaseUrl = "http://image.tmdb.org/t/p/";
-        const string SmallPosterSize = "w185";
-        const string BigPosterSize = "w500";
+        public const int PageSize = 20;
 
+        private const string ApiKey = "1f54bd990f1cdfb230adb312546d765d";
+        protected const string Language = "en-US";
 
-        private const string apiKey = "1f54bd990f1cdfb230adb312546d765d";
-        private IApiService movieApiService;
-        private ICache movieCache;
+        private IApiService _movieApiService;
+        private ICache _movieCache;
 
         public MovieService(IApiService apiService = null, ICache cache = null)
         {
-            movieApiService = apiService ?? Locator.Current.GetService<IApiService>();
-            movieCache = cache ?? Locator.Current.GetService<ICache>();
+            _movieApiService = apiService ?? Locator.Current.GetService<IApiService>();
+            _movieCache = cache ?? Locator.Current.GetService<ICache>();
+            _internalSourceCache = new SourceCache<Movie, int>(o => o.Id);
         }
 
-        public IObservable<IEnumerable<Movie>> GetUpcomingMovies(int index)
+        public IObservable<Unit> LoadUpcomingMovies(int index)
         {
-            return
-                movieCache
-                    .GetAndFetchLatest($"upcoming_movies_{index}", () => FetchUpcomingMovies(index));
+            return _movieCache
+                .GetAndFetchLatest($"upcoming_movies_{index}", () => FetchUpcomingMovies(index))
+                .Select(x =>
+                {
+                    _internalSourceCache.Edit(innerCache => innerCache.AddOrUpdate(x));
+                    return Unit.Default;
+                });
         }
 
         IObservable<IEnumerable<Movie>> FetchUpcomingMovies(int index)
         {
+            // Uncomment this if you want to see the dialog 
+            //var randomNumer = new Random(DateTime.Now.Millisecond).Next(0, 3);
+            //if (randomNumer == 1)
+                //throw new Exception("This is a generic fake exception to show how the dialog works.");
+
             int page = (int)Math.Ceiling(index / (double)PageSize) + 1;
 
             return Observable
                 .CombineLatest(
-                    movieApiService
+                    _movieApiService
```

Comment:
I think it's worth to use extension method instead of the static one. It would look like this:

```csharp
return _movieApiService
                .UserInitiated
                .FetchUpcomingMovies(ApiKey, page, Language)
                .CombineLatest(_movieApiService
                        .UserInitiated
                        .FetchGenres(ApiKey, Language),
                    (movies, genres) =>
                    {
                        return movies
                            .Results
                            .Select(movieDto => MovieMapper.ToModel(genres, movieDto, Language));
                    });
```